### PR TITLE
[JIT] Get rid of emitApplyIdent

### PR DESCRIPTION
**Pull request by jamesr66a**
And reroute builtin/CompilationUnit function resolution through one resolution pathway

### CODE REVIEW

File: test/cpp/jit/tests.h [hunk 1]
```diff
@@ -849,7 +849,7 @@ const auto cf_examples = R"JIT(
 void testControlFlow() {
   script::Module cu;
   script::defineMethodsInModule(
-      cu, cf_examples, torch::jit::script::nativeResolver, nullptr);
+      cu, cf_examples, std::shared_ptr<torch::jit::script::NativeResolver>(), nullptr);
```
Comment (apaszke) [inline test/cpp/jit/tests.h]: Why not pass a `nullptr` if that's what you wanted to do (unless you really wanted `make_shared`)?

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -152,11 +152,11 @@ struct CastValue : public SugaredValue {
 //      the IR API, but for now we choose to pessimisitically create inputs and
 //      delete unnecessary ones later with replaceAllusesWith().
 struct Environment {
-  Environment(Method & method, const Resolver& resolver, Block* b, std::shared_ptr<Environment> next = nullptr)
+  Environment(Method & method, std::shared_ptr<Resolver> resolver, Block* b, std::shared_ptr<Environment> next = nullptr)
       : method(method), resolver(resolver), b(b), next(next) {}
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: nit: move the resolver in

**Review by apaszke — APPROVED** [review]
Two minor comments

File: test/cpp/jit/tests.h (hunk 1 above)
Comment (jamesr66a) [reply test/cpp/jit/tests.h]: Because this was a regex find-replace :D I'll fix it

**Comment by alsrgv** [comment]
@jamesr66a, @apaszke, this dependency on `builtins` causes issues if `future` is not installed.  Can you add it to setup depedencies?

```
Successfully installed torch-nightly-1.0.0.dev20181017
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-5B_sEo'
root@madlj01-sjc1:~# python
Python 2.7.9 (default, Sep 25 2018, 20:42:16)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/torch/__init__.py", line 291, in <module>
    import torch.jit
  File "/usr/local/lib/python2.7/dist-packages/torch/jit/__init__.py", line 10, in <module>
    import builtins
ImportError: No module named builtins
>>>
root@madlj01-sjc1:~# pip install future
Collecting future
  Downloading https://files.pythonhosted.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz (824kB)
    100% |################################| 829kB 12.5MB/s
Building wheels for collected packages: future
  Running setup.py bdist_wheel for future ... done
  Stored in directory: /root/.cache/pip/wheels/bf/c9/a3/c538d90ef17cf7823fa51fc701a7a7a910a80f6a405bf15b1a
Successfully built future
Installing collected packages: future
Successfully installed future-0.16.0
root@madlj01-sjc1:~# python
Python 2.7.9 (default, Sep 25 2018, 20:42:16)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>>
```

**Comment by jamesr66a** [comment]
@alsrgv if you replace `builtin` with `__builtin__` in the source file does it work? I think we can conditionally import one or the other

**Comment by jamesr66a** [comment]
@alsrgv  in particular i want to see if this is a solution https://github.com/pytorch/pytorch/pull/12784

**Comment by apaszke** [comment]
The extra dependency is a bug. We should remove it!

**Comment by jamesr66a** [comment]
@apaszke the extra dependency is not a bug because we need to match builtins like `len` and those appear in neither `globals` nor `locals`. https://github.com/pytorch/pytorch/pull/12784 should fix the issue

**Comment by apaszke** [comment]
Well adding a dependency on `future` is a bug, but good to hear that it's resolved!

**Pull request closed without merging**